Framework/PythonInterface/test/python/plugins/algorithms/GenerateLogbookTest.py

```diff
@@ -0,0 +1,83 @@
+#   NScD Oak Ridge National Laboratory, European Spallation Source,
+#   Institut Laue - Langevin & CSNS, Institute of High Energy Physics, CAS
+# SPDX - License - Identifier: GPL - 3.0 +
+import unittest
+from mantid.api import mtd, ITableWorkspace
+from mantid.simpleapi import config, GenerateLogbook
+import os
+from tempfile import gettempdir
+
+class GenerateLogbookTest(unittest.TestCase):
+
+    _data_directory = None
+
+    def setUp(self):
+        data_dirs = config['datasearch.directories'].split(';')
+        unit_test_data_dir = [p for p in data_dirs if 'UnitTest' in p][0]
+        d7_dir = 'ILL/D7'
+        if 'ILL' in unit_test_data_dir:
+            d7_dir = 'D7'
+        self._data_directory = os.path.abspath(os.path.join(unit_test_data_dir,  d7_dir))
+
+    def tearDown(self):
+        mtd.clear()
+
+    @classmethod
+    def tearDownClass(cls):
+        mtd.clear()
+        if os.path.exists(os.path.join(gettempdir(), 'logbook.csv')):
+            os.remove(os.path.join(gettempdir(), 'logbook.csv'))
+
+    def test_instrument_does_not_exist(self):
+        self.assertTrue(os.path.exists(self._data_directory))
+        with self.assertRaises(RuntimeError):
+            GenerateLogbook(Directory=self._data_directory,
+                            OutputWorkspace='__unused', Facility='ISIS', Instrument='nonexistent')
+
+    def test_d7_default(self):
+        self.assertTrue(os.path.exists(self._data_directory))
+        GenerateLogbook(Directory=self._data_directory,
+                        OutputWorkspace='default_logbook', Facility='ILL', Instrument='D7',
+                        NumorRange=[396990,396993])
+        self._check_output('default_logbook', numberEntries=3, numberColumns=6)
+
+    def test_d7_optional(self):
+        self.assertTrue(os.path.exists(self._data_directory))
+        GenerateLogbook(Directory=self._data_directory,
+                        OutputWorkspace='optional_logbook', Facility='ILL', Instrument='D7',
+                        NumorRange=[396990,396993], OptionalHeaders='TOF')
+        self._check_output('optional_logbook', numberEntries=3, numberColumns=7)
+
+    def test_d7_custom(self):
+        self.assertTrue(os.path.exists(self._data_directory))
+        GenerateLogbook(Directory=self._data_directory,
+                        OutputWorkspace='custom_logbook', Facility='ILL', Instrument='D7',
+                        NumorRange=[396990,396993], CustomEntries='/entry0/acquisition_mode')
+        self._check_output('custom_logbook', numberEntries=3, numberColumns=7)
+
+    def test_d7_custom_with_summing(self):
+        self.assertTrue(os.path.exists(self._data_directory))
+        GenerateLogbook(Directory=self._data_directory,
+                        OutputWorkspace='custom_logbook_w_summing', Facility='ILL', Instrument='D7',
+                        NumorRange=[396990,396993], CustomEntries='/entry0/acquisition_mode,polarisation')
```

Comment:
I'm not sure this test actually triggers the use of a binary operation (+). It looks like you've either got to put the formula for polarisation in here:
```
CustomEntries='/entry0/acquisition_mode, /entry0/D7/POL/actual_state+/entry0/D7/POL/actual_stateB1B2'
```
or you need to specify polarisation in the OptionalHeaders parameter. Please can you make one of these changes so the test executes the binary operation code in GenerateLogbook.py?

If do the first approach then it seems like the code to determine the column header in the output table workspace doesn't work 100% correctly in that it decides the column is called actual_stateB1B2 (just strips off the final string). This point about the column header not a big deal though.

Reply:
Yes indeed, this is not correct. The OptionalHeaders need to be defined for this operation to be performed. The original operation would search for a single entry of the provided structure, which was not possible to be found. This test is fixed and also the name creation has been updated to be derived from the entries going into the operation to be performed.

Reply:
Maybe you've not finished work on this yet, but I don't think this unit test has been updated? I can see you've updated the column header for entries that use a formula thanks.

Reply:
My apologies, after adding this correction to the doc test, I missed changing the unit test. I've just corrected it.